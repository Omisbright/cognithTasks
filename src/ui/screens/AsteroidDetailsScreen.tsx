import React, {Component} from 'react';
import {View, Text} from 'react-native';
import {AppProps} from '../../navigation/types';
import {asteroidDetailsScreenStyles} from '../styles/styles';
import axios from 'axios';

type State = {
  asteroidData: {
    name: string;
    estimated_diameter: {
      kilometers: {
        estimated_diameter_min: number;
        estimated_diameter_max: number;
      };
    };
    is_potentially_hazardous_asteroid: boolean;
  } | null;
  loading: boolean;
};

class AsteroidDetailsScreen extends Component<
  AppProps<'AsteroidDetailsScreen'>,
  State
> {
  constructor(props: AppProps<'AsteroidDetailsScreen'>) {
    super(props);
    this.state = {
      asteroidData: null,
      loading: false,
    };
  }

  componentDidMount() {
    const asteroidId = this.props.route.params?.asteroidId;
    this.fetchAsteroidDetails(asteroidId);
  }

  fetchAsteroidDetails(id: string | undefined) {
    this.setState({loading: true});
    const apiKey = 'eGo5f1A3A8Z6xlOZW1fbLBuVEO0vgJIdhWAQrjH8';
    const baseUrl = 'https://api.nasa.gov/neo/rest/v1/neo/';
    let asteroidUrl = '';

    if (id) {
      asteroidUrl = `${baseUrl}${id}?api_key=${apiKey}`;
      this.fetchAsteroidData(asteroidUrl);
    } else {
      const randomAsteroidUrl = `${baseUrl}browse?api_key=${apiKey}`;
      axios
        .get(randomAsteroidUrl)
        .then(response => {
          this.setState({loading: false});
          const data = response.data;
          const randomIndex = Math.floor(
            Math.random() * data.near_earth_objects.length,
          );
          const randomAsteroid = data.near_earth_objects[randomIndex];
          asteroidUrl = `${baseUrl}${randomAsteroid.id}?api_key=${apiKey}`;
          this.fetchAsteroidData(asteroidUrl);
        })
        .catch(error => {
          this.setState({loading: false});
          console.log('error', error);
        });
    }
  }

  fetchAsteroidData(asteroidUrl: string) {
    this.setState({loading: true});
    axios
      .get(asteroidUrl)
      .then(asteroidResponse => {
        this.setState({loading: false});
        const asteroidData = asteroidResponse.data;
        this.setState({asteroidData});
      })
      .catch(error => {
        this.setState({loading: false});
        console.log('error', error);
      });
  }

  render() {
    const {asteroidData, loading} = this.state;

    if (loading) {
      return (
        <View style={asteroidDetailsScreenStyles.container}>
          <Text>Loading...</Text>
        </View>
      );
    }

    if (!asteroidData) {
      return (
        <View style={asteroidDetailsScreenStyles.container}>
          <Text>No asteroid data found</Text>
        </View>
      );
    }

    return (
      <View style={asteroidDetailsScreenStyles.container}>
        <Text>Name: {asteroidData.name}</Text>
        <Text>
          Estimated Diameter:{' '}
          {asteroidData.estimated_diameter.kilometers.estimated_diameter_min} -{' '}
          {asteroidData.estimated_diameter.kilometers.estimated_diameter_max} km
        </Text>
        <Text>
          Potentially Hazardous:{' '}
          {asteroidData.is_potentially_hazardous_asteroid ? 'Yes' : 'No'}
        </Text>
      </View>
    );
  }
}

export default AsteroidDetailsScreen;
