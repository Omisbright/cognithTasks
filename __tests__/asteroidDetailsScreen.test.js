import React from 'react';
import {render, waitFor} from '@testing-library/react-native';
import AsteroidDetailsScreen from '../src/ui/screens/AsteroidDetailsScreen';
import axios from 'axios';
import renderer from 'react-test-renderer';

jest.mock('axios');

describe('AsteroidDetailsScreen', () => {
  const asteroidData = {
    name: 'Test Asteroid',
    estimated_diameter: {
      kilometers: {
        estimated_diameter_min: 1,
        estimated_diameter_max: 2,
      },
    },
    is_potentially_hazardous_asteroid: true,
  };

  beforeEach(() => {
    axios.get.mockClear();
  });

  it('renders correctly', async () => {
    const mockResponse = {
      data: {
        url: '2000405',
      },
    };

    axios.get.mockResolvedValue(mockResponse);
    const tree = renderer
      .create(<AsteroidDetailsScreen route={{params: {asteroidId: '123'}}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render asteroid details when mounted with asteroidId', async () => {
    axios.get.mockResolvedValueOnce({data: asteroidData});

    const {getByText} = render(
      <AsteroidDetailsScreen route={{params: {asteroidId: '123'}}} />,
    );

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(getByText(`Name: ${asteroidData.name}`)).toBeTruthy();
    expect(
      getByText(
        `Estimated Diameter: ${asteroidData.estimated_diameter.kilometers.estimated_diameter_min} - ${asteroidData.estimated_diameter.kilometers.estimated_diameter_max} km`,
      ),
    ).toBeTruthy();
    expect(
      getByText(
        `Potentially Hazardous: ${
          asteroidData.is_potentially_hazardous_asteroid ? 'Yes' : 'No'
        }`,
      ),
    ).toBeTruthy();
  });

  it('should render asteroid details when mounted without asteroidId', async () => {
    const route = {
      params: {asteroidId: undefined}, // Provide the necessary params here
    };

    const randomAsteroid = {
      id: '456',
    };

    axios.get
      .mockResolvedValueOnce({data: {near_earth_objects: [randomAsteroid]}})
      .mockResolvedValueOnce({data: asteroidData});

    const {getByText} = render(<AsteroidDetailsScreen route={route} />);

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(2));

    expect(getByText(`Name: ${asteroidData.name}`)).toBeTruthy();
    expect(
      getByText(
        `Estimated Diameter: ${asteroidData.estimated_diameter.kilometers.estimated_diameter_min} - ${asteroidData.estimated_diameter.kilometers.estimated_diameter_max} km`,
      ),
    ).toBeTruthy();
    expect(
      getByText(
        `Potentially Hazardous: ${
          asteroidData.is_potentially_hazardous_asteroid ? 'Yes' : 'No'
        }`,
      ),
    ).toBeTruthy();
  });

  it('should render error message when API call fails', async () => {
    const route = {
      params: {asteroidId: undefined}, // Provide the necessary params here
    };

    axios.get.mockRejectedValueOnce(new Error('API Error'));

    const {getByText} = render(<AsteroidDetailsScreen route={route} />);

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(getByText('No asteroid data found')).toBeTruthy();
  });
});
