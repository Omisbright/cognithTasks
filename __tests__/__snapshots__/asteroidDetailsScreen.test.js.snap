// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsteroidDetailsScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Loading...
  </Text>
</View>
`;
